### PR TITLE
Add MiqPassword.encrypt_hash & decrypt_hash.

### PR DESCRIPTION
**Pull request by lfu**
These methods are used for the password fields in options of ServiceTemplate and Service.
 
https://www.pivotaltracker.com/n/projects/1937537/stories/139479571

Part of #https://github.com/ManageIQ/manageiq/pull/13934

cc @gmcculloug @bzwei

### CODE REVIEW

File: spec/util/miq-password_spec.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 require 'util/miq-password'
 require 'tempfile'
+require 'util/extensions/miq-deep'
```
Comment (carbonin) [inline spec/util/miq-password_spec.rb]: I feel like this should be included in `miq-password.rb` as that's where `deep_clone` is being used.

File: spec/util/miq-password_spec.rb [hunk 2]
```diff
@@ -353,6 +354,54 @@
...
+      expect(new["my_password"]).to be_empty
+      expect(new.keys.include?("password::my_password")).to be false
+    end
+
+    it "should add 'password::' prefix" do
+      options = { "my_password" => "secret" }
+      new = MiqPassword.encrypt_hash(options)
+      expect(new.keys.first).to eq("password::my_password")
+      expect(new["password::my_password"]).to be_encrypted("secret")
+      expect(new.keys.include?("my_password")).to be false
```
Comment (carbonin) [inline spec/util/miq-password_spec.rb]: Please don't use keywords as variable names, it makes things like this ambiguous.

Comment (lfu) [inline spec/util/miq-password_spec.rb]: Thought there was something unusual 👍

File: lib/gems/pending/util/miq-password.rb [hunk 3]
```diff
@@ -72,6 +73,40 @@ def self.encrypted?(str)
...
   end
 
+  def self.encrypt_hash(values)
+    hash_values = values.deep_clone
+
+    hash_values.select { |f, _| f =~ /[Pp]assword/ }.each do |key, val|
+      next if val.strip.empty?
+
+      key_in_symbol = key.kind_of?(Symbol)
+      hash_values.delete(key)
```
Comment (bzwei) [inline lib/gems/pending/util/miq-password.rb]: can you double check if it is safe to delete/insert an hash entry within an iteration?

File: lib/gems/pending/util/miq-password.rb [hunk 4]
```diff
@@ -72,6 +73,40 @@ def self.encrypted?(str)
...
+
+      key = "password::#{key}" unless key =~ /password::/
```
Comment (bzwei) [inline lib/gems/pending/util/miq-password.rb]: Need to rethink the scenario where key already starts with `password::` - here the key name is not changed, but the value is encrypted. It is no longer symmetric. Then the `#decrpyt_hash` method will not get back exactly the same

Comment (carbonin) [inline lib/gems/pending/util/miq-password.rb]: I'm not actually a huge fan of this in general. `try_decrypt` already checks to see if the thing you gave it is encrypted so it doesn't feel like there should be a reason to use this to indicate that something is a password field. It feels like keeping the key regex's the same would be enough.

File: lib/gems/pending/util/miq-password.rb [hunk 5]
```diff
@@ -72,6 +73,40 @@ def self.encrypted?(str)
     !!split(str).first
   end
 
+  def self.encrypt_hash(values)
+    hash_values = values.deep_clone
+
+    hash_values.select { |f, _| f =~ /[Pp]assword/ }.each do |key, val|
```
Comment (carbonin) [inline lib/gems/pending/util/miq-password.rb]: Regarding https://github.com/ManageIQ/manageiq-gems-pending/pull/63/files#r101402312 I think this regex should be put into a constant or method and used by both `encrypt_hash` and `decrypt_hash`.

File: lib/gems/pending/util/miq-password.rb (hunk 4 above)
Comment (kbrock) [inline lib/gems/pending/util/miq-password.rb]: This is a slippery slope, knowing what is a password.

Wonder if it makes sense to look into `Vmdb::Settings::Walker` - possibly moving that walker into here

File: lib/gems/pending/util/miq-password.rb (hunk 5 above)
Comment (kbrock) [inline lib/gems/pending/util/miq-password.rb]: Please don't hardcode the name of passwords in here.

File: lib/gems/pending/util/miq-password.rb [hunk 6]
```diff
@@ -72,6 +73,28 @@ def self.encrypted?(str)
     !!split(str).first
   end
 
+  def self.encrypt_hash(values, &block)
```
Comment (bzwei) [inline lib/gems/pending/util/miq-password.rb]: Please add comments on what do you expect from the block - the purpose and the return type.

Comment (kbrock) [inline lib/gems/pending/util/miq-password.rb]: no - please don't enhance `encrypt_hash` - please use our existing mechanisms.

**Comment by Fryguy** [comment]
I second the sentiments of @kbrock and @carbonin .  This feels like the wrong solution.  Hash walking is not the domain of MiqPassword.

**Pull request closed without merging**